JIT: Extend escape analysis to account for arrays with non-gcref elements

hez2010 opened the pull request:
Positive case:

```cs
var chs = new char[42];
chs[1] = 'a';
Console.WriteLine((int)chs[1] + chs.Length);
```

Codegen:

```asm
; Assembly listing for method ArrayAllocator.Program:Main() (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; FullOpts code
; optimized code
; rsp based frame
; partially interruptible
; No PGO data
; Final local variable assignments
;
;* V00 loc0         [V00    ] (  0,  0   )    long  ->  zero-ref    class-hnd exact <short[]>
;  V01 OutArgs      [V01    ] (  1,  1   )  struct (32) [rsp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;* V02 tmp1         [V02    ] (  0,  0   )  struct (104) zero-ref    do-not-enreg[SF] "stack allocated array temp"
;* V03 tmp2         [V03    ] (  0,  0   )    long  ->  zero-ref    single-def "V02.[000..008)"
;* V04 tmp3         [V04    ] (  0,  0   )     int  ->  zero-ref    single-def "V02.[008..012)"
;* V05 tmp4         [V05    ] (  0,  0   )   short  ->  zero-ref    "V02.[018..020)"
;
; Lcl frame size = 40

G_M25548_IG01:  ;; offset=0x0000
       sub      rsp, 40
                                                ;; size=4 bbWeight=1 PerfScore 0.25
G_M25548_IG02:  ;; offset=0x0004
       mov      ecx, 84
       call     [System.Console:WriteLine(int)]
       nop
                                                ;; size=12 bbWeight=1 PerfScore 3.50
G_M25548_IG03:  ;; offset=0x0010
       add      rsp, 40
       ret
                                                ;; size=5 bbWeight=1 PerfScore 1.25

; Total bytes of code 21, prolog size 4, PerfScore 5.00, instruction count 6, allocated bytes for code 21 (MethodHash=5b0b9c33) for method ArrayAllocator.Program:Main() (FullOpts)
```

Negative case:

```cs
var chs = new char[42];
chs[1] = 'a';
Console.WriteLine((int)chs[42] + chs.Length);
```

Codegen:

```asm
; Assembly listing for method ArrayAllocator.Program:Main() (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; FullOpts code
; optimized code
; rsp based frame
; partially interruptible
; No PGO data
; Final local variable assignments
;
;* V00 loc0         [V00    ] (  0,  0   )    long  ->  zero-ref    class-hnd exact <short[]>
;  V01 OutArgs      [V01    ] (  1,  1   )  struct (32) [rsp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;* V02 tmp1         [V02    ] (  0,  0   )  struct (104) zero-ref    do-not-enreg[SF] "stack allocated array temp"
;  V03 tmp2         [V03,T00] (  1,  0   )   byref  ->  rbx         must-init "dummy temp of must thrown exception"
;* V04 tmp3         [V04    ] (  0,  0   )    long  ->  zero-ref    single-def "V02.[000..008)"
;* V05 tmp4         [V05    ] (  0,  0   )     int  ->  zero-ref    single-def "V02.[008..012)"
;* V06 tmp5         [V06    ] (  0,  0   )   short  ->  zero-ref    single-def "V02.[018..020)"
;
; Lcl frame size = 32

G_M25548_IG01:  ;; offset=0x0000
       push     rbx
       sub      rsp, 32
       xor      ebx, ebx
                                                ;; size=7 bbWeight=0 PerfScore 0.00
G_M25548_IG02:  ;; offset=0x0007
       call     CORINFO_HELP_RNGCHKFAIL
       movsx    rcx, word  ptr [rbx]
       call     [System.Console:WriteLine(int)]
       int3
                                                ;; size=16 bbWeight=0 PerfScore 0.00

; Total bytes of code 23, prolog size 5, PerfScore 0.00, instruction count 7, allocated bytes for code 23 (MethodHash=5b0b9c33) for method ArrayAllocator.Program:Main() (FullOpts)
; ============================================================
```

Benchmark on Mandelbrot: 

| Method     | Job                    | Mean     | Error   | StdDev  | Code Size | Allocated |
|----------- |----------------------- |---------:|--------:|--------:|----------:|----------:|
| MandelBrot | NoStackAllocationArray | 199.7 us | 1.30 us | 1.22 us |   1,996 B |   2.49 KB |
| MandelBrot | StackAllocationArray   | 195.8 us | 1.16 us | 1.08 us |   2,414 B |   1.14 KB |

Diff: https://www.diffchecker.com/bNP4qHdF/

hez2010 committed:
initial prototype

hez2010 committed:
Morph ARR_LENGTH and INDEX_ADDR

hez2010 committed:
Fix incorrect array length storage

hez2010 committed:
Use offset and correct type

hez2010 committed:
handle reassignment

hez2010 committed:
range check

hez2010 committed:
throw range check failure

hez2010 committed:
update comments

hez2010 committed:
add metrics

jakobbotsch commented on src/coreclr/jit/objectalloc.h:
```diff
@@ -33,7 +34,9 @@ class ObjectAllocator final : public Phase
     // definitely-stack-pointing pointers. All definitely-stack-pointing pointers are in both sets.
     BitVec              m_PossiblyStackPointingPointers;
     BitVec              m_DefinitelyStackPointingPointers;
+    BitVec              m_PointersHasLocalStore;
```
Why does the analysis need to change for arrays? That seems odd.

hez2010 replied:
For example, the below code:
```cs
var a = new int[42];
a = new int[43];
```
will be imported to
```
***** BB01 [0000]
STMT00000 ( 0x000[E-] ... 0x007 )
               [000003] DACXG------                         *  STORE_LCL_VAR ref    V00 loc0
               [000002] --CXG------                         \--*  CALL help ref    CORINFO_HELP_NEWARR_1_VC
               [000001] H---------- arg0                       +--*  CNS_INT(h) long   0x7ffb1727d7b0 class int[]
               [000000] ----------- arg1                       \--*  CNS_INT   long   42

***** BB01 [0000]
STMT00001 ( 0x008[E-] ... 0x00F )
               [000007] DACXG------                         *  STORE_LCL_VAR ref    V00 loc0
               [000006] --CXG------                         \--*  CALL help ref    CORINFO_HELP_NEWARR_1_VC
               [000005] H---------- arg0                       +--*  CNS_INT(h) long   0x7ffb1727d7b0 class int[]
               [000004] ----------- arg1                       \--*  CNS_INT   long   43
```
where `V00` gets reused.

This can lead problems like:
- stack allocated arrays doesn't get proper zeroing before re-assignment
- setting it to `null` would lead to an assert in `RewriteUses` at https://github.com/dotnet/runtime/blob/a86987ccab917433d065fe5dc8870fc261f79d14/src/coreclr/jit/objectalloc.cpp#L907

jakobbotsch replied:
Yes, `V00` gets reused, but `V00` is just a pointer to the object. It is fine for it to be pointing to two different stack allocated objects at different times. The same case with a normal object should work just fine.

jakobbotsch replied:
Is this because you are trying to track the array length based on the pointer local (`V00`) here? I think that's the problem -- we should not try to do that. Instead, the IR of `INDEX_ADDR` should be transformed into something that preserves its semantics, yet exposes the potential array access to local morph. We do the same thing (in spirit) for `CORINFO_HELP_UNBOX`, which gets transformed into a `COMMA` that calls `CORINFO_UNBOX_HELPER_TYPETEST` on the LHS.

hez2010 replied:
> Is this because you are trying to track the array length based on the pointer local (V00) here?

No, I'm tracking the array length based on new stack allocated local.

hez2010 replied:
The issue here is that we were using `m_HeapLocalToStackLocalMap` to track the relationship between the original local and the new stack allocated local. When it sees a local which has been stack allocated (ie. a local has the relationship recorded in `m_HeapLocalToStackLocalMap`), it assumes the local must be a use and tries to update it to `LCL_VAR_ADDR`. But for arrays, it can be yet another `STORE_LCL_VAR` instead of a use.

jakobbotsch replied:
So essentially object stack allocation is relying on an IR invariant that all `ALLOCOBJ` nodes must be stored to different locals when we get to object allocation? And also that all uses have the store as its reaching def, and come in the right order in `Compiler::Blocks()`. This seems like a dangerous assumption to make in object stack allocation.

I do not think object stack allocation has a need for this anymore since local morph has proper flow sensitive propagation of these `LCL_ADDR`s now. Seems like something we should fix; the current approach is scary to me.

cc @AndyAyersMS, what do you think?

jakobbotsch replied:
The comment talks about exposing `ALLOCOBJ` as top level statements to be able to find them. That's ok with me; it is an optimization on throughput (though I question it being worth it, compared to just indicating its location with `GTF_CALL` or similar). If we didn't end up finding an `ALLOCOBJ` we just wouldn't stack allocate it.

That's not the only constraint the current approach has. It would also be illegal to optimize `newobj; stloc` into a top-level store directly to that local, or introduce any sort of transformation that does that between import and object stack allocation. If this happened object stack allocation could end up improperly replacing some uses of that local by a `LCL_ADDR`. That's a much more subtle invariant that the comment does not talk about.

AndyAyersMS replied:
I guess I'm not that worried right now, though we should add this to the next steps list (which I need to resurrect).

The phase already asserts that all `ALLOCOBJ` are top level. We could extend that to insist that the store destination is at least marked as single-def temp (which is how the importer arranges things). That would catch some of the substitution issues, though it's not clear how far outside the importer one can trust single-def.

If we want to generalize the analysis or allow different IR shapes we will need to redesign things somewhat.

AndyAyersMS replied:
> I think similar things need to be done for `newarr` as well.

Yes, for now you need to follow this pattern.

jakobbotsch commented on src/coreclr/jit/objectalloc.cpp:
```diff
@@ -400,6 +425,81 @@ bool ObjectAllocator::MorphAllocObjNodes()
+                // Don't attempt to do stack allocations inside basic blocks that may be in a loop.
+                //
+                if (!IsObjectStackAllocationEnabled())
+                {
+                    onHeapReason = "[object stack allocation disabled]";
+                    canStack     = false;
+                }
+                else if (basicBlockHasBackwardJump)
+                {
+                    onHeapReason = "[alloc in loop]";
+                    canStack     = false;
+                }
+                else if (!len->IsCnsIntOrI())
+                {
+                    onHeapReason = "[non-constant size]";
+                    canStack     = false;
+                }
+                else if (!CanAllocateLclVarOnStack(lclNum, clsHnd, (unsigned int)len->AsIntCon()->gtIconVal, &blockSize,
+                                                   &onHeapReason))
+                {
+                    // reason set by the call
+                    canStack = false;
+                }
+                else
+                {
+                    JITDUMP("Allocating V%02u on the stack\n", lclNum);
+                    canStack = true;
+                    const unsigned int stackLclNum =
+                        MorphNewArrNodeIntoStackAlloc(asCall, clsHnd, (unsigned int)len->AsIntCon()->gtIconVal,
+                                                      blockSize, block, stmt);
+                    m_HeapLocalToStackLocalMap.AddOrUpdate(lclNum, stackLclNum);
+                    m_LocalArrToLenMap.AddOrUpdate(stackLclNum, (unsigned int)len->AsIntCon()->gtIconVal);
+                    // We keep the set of possibly-stack-pointing pointers as a superset of the set of
+                    // definitely-stack-pointing pointers. All definitely-stack-pointing pointers are in both sets.
+                    MarkLclVarAsDefinitelyStackPointing(lclNum);
+                    MarkLclVarAsPossiblyStackPointing(lclNum);
+                    stmt->GetRootNode()->gtBashToNOP();
+                    comp->optMethodFlags |= OMF_HAS_OBJSTACKALLOC;
+                    didStackAllocate = true;
+                }
```
Can you factor this duplication?

hez2010 replied:
will refactor this later.

jakobbotsch commented on src/coreclr/jit/objectalloc.cpp:
```diff
@@ -443,7 +543,7 @@ bool ObjectAllocator::MorphAllocObjNodes()
                     canStack     = false;
                 }
-                else if (!CanAllocateLclVarOnStack(lclNum, clsHnd, &onHeapReason))
+                else if (!CanAllocateLclVarOnStack(lclNum, clsHnd, 0, NULL, &onHeapReason))
```
```suggestion
                else if (!CanAllocateLclVarOnStack(lclNum, clsHnd, 0, nullptr, &onHeapReason))
```
Same elsewhere.

hez2010 committed:
minor cleanup

AndyAyersMS reviewed (commented):
For arrays (and also perhaps boxes and ref classes) we ought to have some kind of size limit... possibly similar to the one we use for stackallocs.

We need to be careful we don't allocate a lot of stack for an object that might not be heavily used, as we'll pay per-call prolog zeroing costs.

hez2010 committed:
Introduce new temp and implement local address morphing

hez2010 committed:
handle index out-of-range

hez2010 committed:
Refactor to remove duplicates

hez2010 committed:
Remove invalid asserts

hez2010 committed:
make compiler happy

jakobbotsch commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -826,6 +826,74 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+                EscapeValue(TopValue(0), node);
+                PopValue();
+                break;
+            }
```
Can you move this down in the `switch`? The local cases are much more common.

jakobbotsch commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -826,6 +826,74 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+                assert(TopValue(1).Node() == node->gtGetOp1());
+                assert(TopValue(0).Node() == node->gtGetOp2());
+
+                if (node->gtGetOp1()->OperIs(GT_LCL_VAR, GT_LCL_ADDR) && node->gtGetOp2()->IsCnsIntOrI())
```
```suggestion
                if (node->gtGetOp2()->IsCnsIntOrI())
```

jakobbotsch commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -826,6 +826,74 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+
+                if (node->gtGetOp1()->OperIs(GT_LCL_VAR, GT_LCL_ADDR) && node->gtGetOp2()->IsCnsIntOrI())
+                {
+                    if (TopValue(1).IsAddress() &&
+                        m_compiler->lvaGetDesc(TopValue(1).LclNum())->GetLayout()->IsBlockLayout())
```
```suggestion
                    if (TopValue(1).IsAddress())
```

jakobbotsch commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -826,6 +826,74 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+                        ssize_t offset = node->AsIndexAddr()->gtElemOffset +
+                                         node->gtGetOp2()->AsIntCon()->IconValue() * node->AsIndexAddr()->gtElemSize;
+
+                        if (offset < static_cast<ssize_t>(m_compiler->lvaLclSize(TopValue(1).LclNum())))
```
```suggestion
                        if (offset < static_cast<ssize_t>(m_compiler->lvaLclExactSize(TopValue(1).LclNum())))
```

jakobbotsch commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -826,6 +826,74 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+                        }
+                        else
+                        {
+                            GenTree* gtThrow =
+                                m_compiler->gtNewMustThrowException(CORINFO_HELP_RNGCHKFAIL, node->TypeGet(),
+                                                                    node->AsIndexAddr()->gtStructElemClass);
```
```suggestion
                            GenTree* gtThrow =
                                m_compiler->gtNewMustThrowException(CORINFO_HELP_RNGCHKFAIL, node->TypeGet(),
                                                                    NO_CLASS_HANDLE);
```

jakobbotsch replied:
But better to just replace this entire thing with my below suggestion, there is no reason to allocate a local as `gtNewMustThrowException` does.

jakobbotsch commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -826,6 +826,74 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+                        }
+                        else
+                        {
+                            GenTree* gtThrow =
+                                m_compiler->gtNewMustThrowException(CORINFO_HELP_RNGCHKFAIL, node->TypeGet(),
+                                                                    node->AsIndexAddr()->gtStructElemClass);
+                            m_compiler->lvaGetDesc(gtThrow->AsOp()->gtOp2->AsLclVarCommon())->incLvRefCnt(1, RCS_EARLY);
+                            *use           = gtThrow;
```
```suggestion
                            *use           = gtNewOperNode(GT_COMMA, gtNewHelperCallNode(CORINFO_HELP_RNGCHKFAIL, TYP_VOID), gtNewIconNode(0, TYP_BYREF));
```

jakobbotsch commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -826,6 +826,74 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+                                                                    node->AsIndexAddr()->gtStructElemClass);
+                            m_compiler->lvaGetDesc(gtThrow->AsOp()->gtOp2->AsLclVarCommon())->incLvRefCnt(1, RCS_EARLY);
+                            *use           = gtThrow;
+                            m_stmtModified = true;
```
This doesn't seem like it should fall through.

jakobbotsch commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -826,6 +826,74 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+                assert(TopValue(1).Node() == node);
+                assert(TopValue(0).Node() == node->gtGetOp1());
+
+                if (node->gtGetOp1()->OperIs(GT_LCL_VAR, GT_LCL_ADDR))
```
This should be deleted.

jakobbotsch commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -826,6 +826,74 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+
+                if (node->gtGetOp1()->OperIs(GT_LCL_VAR, GT_LCL_ADDR))
+                {
+                    if (TopValue(0).IsAddress() &&
+                        m_compiler->lvaGetDesc(TopValue(0).LclNum())->GetLayout()->IsBlockLayout())
```
```suggestion
                    if (TopValue(0).IsAddress())
```

jakobbotsch commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -826,6 +826,74 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+                    if (TopValue(0).IsAddress() &&
+                        m_compiler->lvaGetDesc(TopValue(0).LclNum())->GetLayout()->IsBlockLayout())
+                    {
+                        GenTree* gtLclFld =
+                            m_compiler->gtNewLclFldNode(TopValue(0).LclNum(), TYP_INT, OFFSETOF__CORINFO_Array__length);
```
This should account in some way for `TopValue(0).Offset()` -- either assert that it is zero, or add it to the offset.

jakobbotsch commented on src/coreclr/jit/objectalloc.cpp:
```diff
@@ -390,46 +390,32 @@ bool ObjectAllocator::MorphAllocObjNodes()
-                    canonicalAllocObjFound = true;
+                    allocType = OAT_NEWOBJ;
+                }
+                else if (data->IsHelperCall())
+                {
+                    GenTreeCall* call = data->AsCall();
+                    if (call->GetHelperNum() == CORINFO_HELP_NEWARR_1_VC &&
+                        call->gtArgs.GetArgByIndex(0)->GetNode()->IsCnsIntOrI())
```
```suggestion
                else if (data->IsHelperCall(comp, CORINFO_HELP_NEWARR_1_VC ))
                {
                    GenTreeCall* call = data->AsCall();
                    if (call->gtArgs.GetArgByIndex(0)->GetNode()->IsCnsIntOrI())
```

jakobbotsch commented on src/coreclr/jit/objectalloc.cpp:
```diff
@@ -443,51 +429,129 @@ bool ObjectAllocator::MorphAllocObjNodes()
+                        //       \--*  CNS_INT long
+                        //------------------------------------------------------------------------
+
+                        CallArg*             arg       = data->AsCall()->gtArgs.GetArgByIndex(0);
+                        GenTree*             node      = arg->GetNode();
+                        CORINFO_CLASS_HANDLE clsHnd    = (CORINFO_CLASS_HANDLE)node->AsIntConCommon()->IntegralValue();
```
I think this should use `gtGetHelperCallClassHandle` instead.

jakobbotsch commented on src/coreclr/jit/objectalloc.cpp:
```diff
@@ -555,6 +619,78 @@ GenTree* ObjectAllocator::MorphAllocObjNodeIntoHelperCall(GenTreeAllocObj* alloc
     return helperCall;
 }
 
+unsigned int ObjectAllocator::MorphNewArrNodeIntoStackAlloc(GenTreeCall*         newArr,
+                                                            CORINFO_CLASS_HANDLE clsHnd,
+                                                            unsigned int         length,
+                                                            unsigned int         blockSize,
+                                                            BasicBlock*          block,
+                                                            Statement*           stmt)
```
Can you add a function header?

jakobbotsch commented on src/coreclr/jit/objectalloc.cpp:
```diff
@@ -682,6 +818,8 @@ bool ObjectAllocator::CanLclVarEscapeViaParentStack(ArrayStack<GenTree*>* parent
             case GT_NE:
             case GT_NULLCHECK:
+            case GT_ARR_LENGTH:
+            case GT_INDEX_ADDR:
```
```suggestion
```

I do not think we will see a local we are interested in used as an index.

jakobbotsch replied:
Actually, we are of course interested in the array (for some reason I was only thinking about the index) -- but the value of the array local by definition escapes if `GT_INDEX_ADDR` escapes. I think this should be handled in the same way as `GT_ADD` or `GT_FIELD_ADDR` are handled.

jakobbotsch replied:
You should add a test that fails with the current approach (e.g. something that returns the address of an element of a newly allocated array).

hez2010 replied:
> I think this should be handled in the same way as GT_ADD or GT_FIELD_ADDR are handled.

Resolved.

> You should add a test that fails with the current approach

I will add test cases later.

jakobbotsch commented on src/coreclr/jit/objectalloc.h:
```diff
@@ -150,6 +165,25 @@ inline bool ObjectAllocator::CanAllocateLclVarOnStack(unsigned int         lclNu
     }
+    else if (comp->info.compCompHnd->isSDArray(clsHnd))
+    {
+        CorInfoType type = comp->info.compCompHnd->getChildType(clsHnd, &clsHnd);
+        if (type != CORINFO_TYPE_UNDEF && clsHnd == nullptr)
+        {
+            // This is an array of primitive types
+            classSize = (unsigned int)OFFSETOF__CORINFO_Array__data + genTypeSize(JITtype2varType(type)) * length;
+        }
+        else if (comp->info.compCompHnd->isValueClass(clsHnd))
+        {
+            classSize =
+                (unsigned int)OFFSETOF__CORINFO_Array__data + comp->info.compCompHnd->getClassSize(clsHnd) * length;
+        }
+        else
+        {
+            *reason = "[array contains gc refs]";
+            return false;
+        }
```
```suggestion
        CORINFO_CLASS_HANDLE elemClsHnd = NO_CLASS_HANDLE;
        CorInfoType corType = comp->info.compCompHnd->getChildType(clsHnd, &elemClsHnd);
        var_types type = JITtype2varType(corType);
        ClassLayout* elemLayout = type == TYP_STRUCT ? comp->typGetObjLayout(elemClsHnd) : nullptr;
        if (varTypeIsGC(type) || ((elemLayout != nullptr) && elemLayout->HasGCPtr()))
        {
            *reason = "[array contains gc refs]";
            return false;
        }
        
        unsigned elemSize = elemLayout != nullptr ? elemLayout->GetSize() : genTypeSize(type);
        classSize = (unsigned int)OFFSETOF__CORINFO_Array__data + elemSize * length;
```

hez2010 committed:
Address review feedbacks

hez2010 committed:
Fix INDEX_ADDR and add Sub

hez2010 committed:
Support IsAddressLessThan and its friends

hez2010 committed:
Fix assertions

jakobbotsch commented on src/coreclr/jit/objectalloc.cpp:
```diff
@@ -400,14 +400,10 @@ bool ObjectAllocator::MorphAllocObjNodes()
                     allocType = OAT_NEWOBJ;
                 }
-                else if (data->IsHelperCall())
+                else if (data->IsCall() && data->AsCall()->IsHelperCall(comp, CORINFO_HELP_NEWARR_1_VC) &&
```
```suggestion
                else if (data->IsHelperCall(comp, CORINFO_HELP_NEWARR_1_VC) &&
```

(note the overload is new with #104815 so you may need to merge from main)

hez2010 replied:
Done.

hez2010 committed:
Merge remote-tracking branch 'origin/main' into value-array-stack-alloc

hez2010 committed:
Use new overload

hez2010 commented:
@MihuBot

hez2010 committed:
Expose jitconfig

jakobbotsch committed:
JIT: Remove GTF_IND_INVARIANT and GTF_IND_NONFAULTING flags checking

These flags are strictly optimizations. Having them required to be set
for certain indirs based on context of the operand introduces IR
invariants that are annoying to work with since it suddenly becomes
necessary for all transformations to consider "did we just introduce
this IR shape?". Instead, handle these patterns during morph as the
optimization it is and remove the strict flags checking from
`fgDebugCheckFlags`.

jakobbotsch committed:
Remove old comment

jakobbotsch committed:
Remove another assert

jakobbotsch committed:
Count

jakobbotsch committed:
Try 2 at counting

hez2010 committed:
Introduce BBF_HAS_NEWARR

hez2010 committed:
Early exit on debug as well

hez2010 committed:
Update computed flags

hez2010 committed:
Merge remote-tracking branch 'jakobbotsch/indir-flags' into value-array-stack-alloc

hez2010 committed:
Merge remote-tracking branch 'origin/main' into value-array-stack-alloc

hez2010 committed:
Partially revert 39d1ad9

hez2010 committed:
Reuse existing comma node

hez2010 commented:
@MihuBot

hez2010 committed:
Respect IsBoundsChecked

hez2010 committed:
Check lowerbound too

hez2010 committed:
Fix assertion take 2

hez2010 commented:
The asm diff doesn't give us real results due to missing context. Previously we didn't ask for `isSDArray` and `getChildType` for any array may be a candidate for stack allocation.

hez2010 committed:
Remove redundant jit-ee calls

hez2010 commented:
End up removing the call to `isSDArray` at all.

hez2010 commented:
@MihuBot

hez2010 committed:
Merge branch 'main' into value-array-stack-alloc

hez2010 commented:
Stats of stack allocated arrays gathered in my local environment:

```
name: number of stack allocated arrays/number of newarrs
aspnet.run.windows.x64.checked: 2/1654
benchmarks.run.windows.x64.checked: 1/2111
benchmarks.run_pgo.windows.x64.checked: 6/1736
benchmarks.run_tiered.windows.x64.checked: 0/167
coreclr_tests.run.windows.x64.checked: 23/42245
libraries.crossgen2.windows.x64.checked: 0/0
libraries.pmi.windows.x64.checked: 0/3599
libraries_tests.windows.x64.checked: 33/9801
libraries_tests_no_tiered_compilation.windows.x64.checked: 77/10957
realworld.run.windows.x64.checked: 0/502
smoke_tests.nativeaot.windows.x64.checked: 0/857
```

hez2010 committed:
Fix assertion again

hez2010 committed:
Check array length

hez2010 committed:
Fix assertion in another way

hez2010 committed:
Unset the flag to avoid unnecessary assert

hez2010 committed:
Merge branch 'main' into value-array-stack-alloc

hez2010 committed:
Add tests

hez2010 commented:
@jakobbotsch Can you take a look on `Assertion failed '((tree->gtDebugFlags & GTF_DEBUG_NODE_MORPHED) == 0) && "ERROR: Already morphed this node!"'`? I cannot figure out a way to fix it properly without extensive changes.
Maybe we can just stop creating RNGCHECKFAIL node for the negative case and leave it as is? Though this would lead to unnecessary spilling as the code will end up throwing.

hez2010 committed:
sigh

hez2010 committed:
Support R2R/NativeAOT

hez2010 committed:
Merge branch 'main' into value-array-stack-alloc

hez2010 committed:
Fix building

hez2010 committed:
cleanup

AndyAyersMS reviewed (commented):
I don't think we need to handle the align8 arrays.

But if you do, if you're on a 32 bit platform you probably need to mark the stack allocation local with `lvStructDoubleAlign`. And run some tests to make sure the right things actually happen.

hez2010 committed:
remove invalid assert

AndyAyersMS commented:
Thank you for working on this. I don't see us taking this for .NET 9 but perhaps we can soon thereafter. I will take a closer look next week sometime.

The stats above are not all that encouraging. I think for both arrays and ref classes we likely need to take a more holistic look at what is blocking us from finding more cases (or if indeed, if there are not more cases to be had). Eugene did this sort of thing when the initial support was implemented and found similarly less-than-encouraging results (which is why we never enabled it), but it is worth taking a fresh look.

Did you put up a PR for the field-wise analysis anywhere? That may be one of the key missing pieces.

hez2010 commented:
> Did you put up a PR for the field-wise analysis anywhere? That may be one of the key missing pieces.

It's on my branch. Will open a draft PR soon to validate CI and see how much benefits we can gain from the field-wise analysis.

hez2010 committed:
double align on 32bit platform

hez2010 committed:
Use correct alignment for align8

hez2010 commented:
@AndyAyersMS @jakobbotsch 

Stats **without** field-wise analysis:

| name | ref-classes (%) | boxes (%) | arrays (%) | arrays once we have gc type support (%) |
| --- | --- | --- | --- | --- |
| benchmarks.run.windows.x64.checked | 53/15090 (0.35%) | 2/1221 (0.16%) | 13/2313 (0.56%) | 16/2313 (0.69%) |
| benchmarks.run_pgo.windows.x64.checked | 6/15777 (0.04%) | 0/1107 (0.00%) | 25/1583 (1.58%) | 26/1583 (1.64%) |
| benchmarks.run_tiered.windows.x64.checked | 2/1857 (0.11%) | 1/179 (0.56%) | 14/236 (5.93%) | 15/236 (6.36%) |
| coreclr_tests.run.windows.x64.checked | 4143/120005 (3.45%) | 15548/86856 (17.90%) | 150/58725 (0.26%) | 305/58725 (0.52%) |
| libraries.crossgen2.windows.x64.checked | 67/105593 (0.06%) | 105/16946 (0.62%) | 13/11330 (0.11%) | 36/11330 (0.32%) |
| libraries.pmi.windows.x64.checked | 361/131612 (0.27%) | 738/18515 (3.99%) | 16/14487 (0.11%) | 48/14487 (0.33%) |
| libraries_tests.run.windows.x64.Release | 289/200745 (0.14%) | 1609/23970 (6.71%) | 89/16790 (0.53%) | 186/16790 (1.11%) |
| libraries_tests_no_tiered_compilation.run.windows.x64.Release | 3899/657627 (0.59%) | 684/120487 (0.57%) | 325/107000 (0.30%) | 756/107000 (0.71%) |
| realworld.run.windows.x64.checked | 31/23481 (0.13%) | 2/1636 (0.12%) | 1/3347 (0.03%) | 5/3347 (0.15%) |
| smoke_tests.nativeaot.windows.x64.checked | 105/9673 (1.09%) | 29/1442 (2.01%) | 0/989 (0.00%) | 8/989 (0.81%) |

Stats **with** field-wise analysis:

| name | ref-classes (%) | boxes (%) | arrays (%) | arrays once we have gc type support (%) |
| --- | --- | --- | --- | --- |
| benchmarks.run.windows.x64.checked | 3/14513 (0.02%) | 2/1208 (0.17%) | 11/2259 (0.49%) | 12/2259 (0.53%) |
| benchmarks.run_pgo.windows.x64.checked | 2/15239 (0.01%) | 0/1105 (0.00%) | 36/1492 (2.41%) | 38/1492 (2.55%) |
| benchmarks.run_tiered.windows.x64.checked | 1/1780 (0.06%) | 1/175 (0.57%) | 10/222 (4.50%) | 10/222 (4.50%) |
| coreclr_tests.run.windows.x64.checked | 3591/117088 (3.07%) | 15493/86660 (17.88%) | 168/58436 (0.29%) | 243/58436 (0.42%) |
| libraries.crossgen2.windows.x64.checked | 8/104655 (0.01%) | 105/16915 (0.62%) | 49/11291 (0.43%) | 62/11291 (0.55%) |
| libraries.pmi.windows.x64.checked | 71/126196 (0.06%) | 735/18323 (4.01%) | 171/14173 (1.21%) | 635/14173 (4.48%) |
| libraries_tests.run.windows.x64.Release | 99/186948 (0.05%) | 1584/23484 (6.75%) | 140/16307 (0.86%) | 362/16307 (2.22%) |
| libraries_tests_no_tiered_compilation.run.windows.x64.Release | 1687/633620 (0.27%) | 659/119704 (0.55%) | 1246/105744 (1.18%) | 1714/105744 (1.62%) |
| realworld.run.windows.x64.checked | 2/22511 (0.01%) | 2/1613 (0.12%) | 3/3232 (0.09%) | 28/3232 (0.87%) |
| smoke_tests.nativeaot.windows.x64.checked | 98/9560 (1.03%) | 29/1442 (2.01%) | 2/987 (0.20%) | 9/987 (0.91%) |

Worth to note that we are missing contexts due to `GetChildType` for some methods, so the real numbers would be more.

hez2010 commented:
Hitting an issue in NativeAOT smoke tests:

```
        Unhandled exception. System.PlatformNotSupportedException: Operation is not supported on this platform.
           at System.Runtime.CompilerServices.RuntimeHelpers.InitializeArray(Array, RuntimeFieldHandle) + 0x3b
           at Program.IsConstantTrue() + 0x6b
           at Program.Check(String, Nullable`1, , Boolean, Func`1) + 0x18d
           at Program.Main() + 0x47e
```

any idea about why `InitializeArray` was not being expanded?

Other failures are unrelated.

hez2010 committed:
Fix intrinsic expansion

hez2010 committed:
Merge branch 'main' into value-array-stack-alloc

hez2010 commented:
> Not sure if it's related to this PR, if yes, any idea about why InitializeArray was not being expanded in this test?

Resolved. Turns out to be that spilling the array to a temp breaks the intrinsic expansion.

hez2010 commented:
Tests (modulus unrelated) are passing. 
PTAL @jakobbotsch 

| name | arrays (%) | arrays with field-wise analysis (%) |
| --- | --- | --- |
| benchmarks.run.windows.x64.checked | 14/4163 (0.34%) | 14/4084 (0.34%) |
| benchmarks.run_pgo.windows.x64.checked | 25/1562 (1.60%) | 36/1550 (2.32%) |
| benchmarks.run_tiered.windows.x64.checked | 14/378 (3.70%) | 10/368 (2.72%) |
| coreclr_tests.run.windows.x64.checked | 162/60525 (0.27%) | 183/60271 (0.30%) |
| libraries.crossgen2.windows.x64.checked | 13/11330 (0.11%) | 48/11300 (0.42%) |
| libraries.pmi.windows.x64.checked | 16/14840 (0.11%) | 325/14577 (2.23%) |
| libraries_tests.run.windows.x64.Release | 79/16584 (0.48%) | 124/16167 (0.77%) |
| libraries_tests_no_tiered_compilation.run.windows.x64.Release | 325/109551 (0.30%) | 1340/108652 (1.23%) |
| realworld.run.windows.x64.checked | 1/3438 (0.03%) | 3/3331 (0.09%) |
| smoke_tests.nativeaot.windows.x64.checked | 0/857 (0.00%) | 2/855 (0.23%) |

hez2010 commented:
I collected the method context against BCL by myself to prevent missing method context.

| name | ref-classes (%) | boxes (%) | arrays (%) |
| --- | --- | --- | --- |
| w/o field-wise | 177/49224 (0.36%) | 380/7015 (5.42%) | 19/5195 (0.37%) |
| with field-wise | 407/55667 (0.73%) | 375/7031 (5.33%) | 38/5823 (0.65%) |

hez2010 commented:
> I don't see us taking this for .NET 9 but perhaps we can soon thereafter. I will take a closer look next week sometime.
> 
> The stats above are not all that encouraging. I think for both arrays and ref classes we likely need to take a more holistic look at what is blocking us from finding more cases (or if indeed, if there are not more cases to be had). Eugene did this sort of thing when the initial support was implemented and found similarly less-than-encouraging results (which is why we never enabled it), but it is worth taking a fresh look.

It's actually a small change as it relies on the existing analysis and added the treatment to arrays as well, which is a "free" improvement. I personally would like to see it goes to .NET 9.

hez2010 committed:
Merge remote-tracking branch 'origin/main' into value-array-stack-alloc

hez2010 committed:
Merge branch 'main' into value-array-stack-alloc

JulieLeeMSFT commented:
@AndyAyersMS please review this community PR.

AndyAyersMS commented on src/coreclr/jit/gentree.h:
```diff
@@ -7534,12 +7534,11 @@ struct GenTreeArrAddr : GenTreeUnOp
         , m_firstElemOffset(firstElemOffset)
     {
-        assert(addr->TypeIs(TYP_BYREF));
         assert(((elemType == TYP_STRUCT) && (elemClassHandle != NO_CLASS_HANDLE)) ||
```
Can you re-add the assert? We only expect TYP_BYREF or TYP_I_IMPL here (for the stack allocated case).

AndyAyersMS commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -1103,6 +1103,75 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+                assert(TopValue(1).Node() == node->gtGetOp1());
+                assert(TopValue(0).Node() == node->gtGetOp2());
+
+                if (node->gtGetOp2()->IsCnsIntOrI() && TopValue(1).IsAddress())
```
Add a comment here that we only expect `TopValue(1).IsAddress()` to be true for stack allocated arrays?

AndyAyersMS commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -1103,6 +1103,75 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+
+                    if (!node->AsIndexAddr()->IsBoundsChecked() ||
+                        (static_cast<ssize_t>(node->AsIndexAddr()->gtElemOffset) <= offset &&
+                         offset < static_cast<ssize_t>(m_compiler->lvaLclExactSize(TopValue(1).LclNum()))))
```
And one here that notes that for stack allocated arrays the local size is the size of the entire array.

AndyAyersMS commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -1103,6 +1103,75 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+                    }
+                    else
+                    {
+                        *use           = m_compiler->gtNewOperNode(GT_COMMA, node->TypeGet(),
```
Do we really need to handle the out of bounds cases here?

hez2010 replied:
IIRC not handing it will make the negative case address exposed, while inserting a BOUND_CHECK node will lead to an assert in later phase.

AndyAyersMS commented on src/coreclr/jit/morph.cpp:
```diff
@@ -8358,10 +8358,20 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac, bool* optA
             return tree;
         }
 
+#ifdef DEBUG
```
As I noted over in #107542 we don't want to have control flow logic conditional on `DEBUG` like this...  we need to find a better fix.

AndyAyersMS reviewed (commented):
Some questions along with the particular notes:
* do you see diffs with these changes when `DOTNET_JitObjectStackAllocationArray=0`?
* are stack allocated arrays amenable to physical promotion? If not, why not?
* what does codegen look like for a locally allocated array consumed by a loop? Is it at least as good as the heap case? 
* we need a tighter size limit, 8192 is too high. We likely need one both for individual allocations (especially if conditional) and for the overall amount of stack size increase. For individual allocations we should either match the stackalloc -> local size limit or gather data suggesting why a different choice is the right one. For the overall limit I'm less sure, but something around 8192 for all stack allocated locals seems about right.
* given that, we might need a prioritization scheme, if we have so many allocation opportunities that some get left out. But I expect this will be rare.

hez2010 commented:
- do you see diffs with these changes when DOTNET_JitObjectStackAllocationArray=0?

I will test it and post the result later.

- are stack allocated arrays amenable to physical promotion? If not, why not?

I think yes? 

```cs
int[] a = new[] { 1,2,3,4 };
return a[2];
```

will produce the following codegen:

```asm
mov rax, 3
ret
```

- what does codegen look like for a locally allocated array consumed by a loop? Is it at least as good as the heap case?

We need https://github.com/dotnet/runtime/pull/102965 to get better codegen for this case. Otherwise it gets address exposed. 

- we need a tighter size limit, 8192 is too high. We likely need one both for individual allocations (especially if conditional) and for the overall amount of stack size increase. For individual allocations we should either match the stackalloc -> local size limit or gather data suggesting why a different choice is the right one. For the overall limit I'm less sure, but something around 8192 for all stack allocated locals seems about right.

Agree. I think we can do this in a following PR?

hez2010 committed:
Merge branch 'main' into value-array-stack-alloc

hez2010 commented:
> do you see diffs with these changes when DOTNET_JitObjectStackAllocationArray=0?

Just checked superpmi and there's no stack allocated array when `DOTNET_JitObjectStackAllocationArray=0`.

hez2010 committed:
Address some review feedback

hez2010 committed:
Merge branch 'main' into value-array-stack-alloc

hez2010 committed:
Revert workaround to fgMorphSmpOp

hez2010 commented:
Test failures are #107542. I'm going to leave it as is.

AndyAyersMS commented:
@hez2010 can you merge this up, then we'll take a fresh look?

hez2010 committed:
Merge branch 'main' into value-array-stack-alloc

hez2010 commented:
> @hez2010 can you merge this up, then we'll take a fresh look?

Done.

hez2010 committed:
Fix build failure

hez2010 committed:
Merge branch 'main' into value-array-stack-alloc

hez2010 committed:
Try fixing remorph issue

AndyAyersMS commented on src/coreclr/jit/objectalloc.cpp:
```diff
@@ -347,17 +347,25 @@ void ObjectAllocator::ComputeStackObjectPointers(BitVecTraits* bitVecTraits)
                         unsigned bitCount = BitVecOps::Count(bitVecTraits, m_ConnGraphAdjacencyMatrix[lclNum]);
-                        assert(bitCount <= 1);
-                        if (bitCount == 1)
+                        if (bitCount > 0)
```
Do you recall why you changed the logic here?

hez2010 replied:
IIRC it was done in 39d1ad9a2ba71031550f92259c1626e0ec497404, to resolve Pri0 test failures.

hez2010 replied:
Let's see how CI is going after bringing back the assertion.

hez2010 replied:
I just checked `b89546`, we end up the following gentree:
```
***** BB02 [0001]
STMT00002 ( 0x004[E-] ... 0x00B )
               [000009] DA-XG------                         *  STORE_LCL_VAR long   V01 loc1
               [000008] ---XG------                         \--*  INDEX_ADDR byref RangeTest[]
               [000006] -----------                            +--*  LCL_VAR   ref    V00 loc0
               [000007] -----------                            \--*  LCL_VAR   long   V02 loc2
```
which points both `V00` and `V02` to `V01`, and causes `bitCount` to be `2`, which seems incorrect.
We should only point `V00` to `V01` in this case instead. I'll push a fix to this.

hez2010 commented:
Resolved `Assertion failed '((tree->gtDebugFlags & GTF_DEBUG_NODE_MORPHED) == 0) && "ERROR: Already morphed this node!"'` in 1914e806485b8b90ad031ca09607874e09e096e1. Maybe worth to check whether this resolves #107542 and #56962 as well.

hez2010 committed:
Minimize asmdiff for tier-0

hez2010 committed:
Check CI

hez2010 committed:
Take INDEX as non-escape

hez2010 committed:
Early exit when optimization is disabled

hez2010 committed:
Format

hez2010 commented:
@AndyAyersMS Now all tests are green, and this is ready for merge, please take another look.
@MihuBot

AndyAyersMS commented:
I have some other changes to escape analysis which are going to conflict, so my plan is to merge those first and then pick this (or something like it) up later. Not sure how long that will take, hopefully not too long.

In the meantime, could you check if your changes to `gtFoldExpr` and morph resolve https://github.com/dotnet/runtime/issues/107542, and if so, split those off separately?

Also if you want to peel off the change to always use a temp for `newarr` we could take that in advance too; it would be nice to see it go in as a zero diff prerequisite.

AndyAyersMS commented:
@hez2010 can you resolve conflicts? The work I was doing was held up so maybe we can work on this and get it in first.

hez2010 committed:
Merge branch 'main' into value-array-stack-alloc

hez2010 committed:
Revert "Try fixing remorph issue"

This reverts commit 1914e806485b8b90ad031ca09607874e09e096e1.

hez2010 commented:
@MihuBot

AndyAyersMS committed:
Merge branch 'main' into value-array-stack-alloc

AndyAyersMS commented:
Hopefully #110787 unblocks this.

AndyAyersMS commented:
Some interesting [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=899067&view=ms.vss-build-web.run-extensions-tab) from SPMI

hez2010 commented:
One of the regression coming from `Array.ForEach(new int[1], null)`:

```diff
-       sub      rsp, 40
+       sub      rsp, 56
 						;; size=4 bbWeight=0 PerfScore 0.00
 G_M52314_IG02:        ; bbWeight=0, gcrefRegs=0000 {}, byrefRegs=0000 {}, byref
+       vxorps   xmm0, xmm0, xmm0
+       vmovdqu  xmmword ptr [rsp+0x20], xmm0
+       vmovdqu  xmmword ptr [rsp+0x24], xmm0
+       mov      rcx, 0xD1FFAB1E      ; int[]
+       mov      qword ptr [rsp+0x20], rcx
+       mov      dword ptr [rsp+0x28], 1
        mov      ecx, 28
        call     [System.ThrowHelper:ThrowArgumentNullException(int)]
        ; gcr arg pop 0
        int3     
```

The loop was originally unrolled but now it's no longer doing that. Seems that we need to propagate the assertion into loops so that the constant size can be propagated into natural loops to make the loop able to be unrolled, which is #110501

AndyAyersMS commented:
I will dig into some of these tomorrow. Need to look closely at the dumps.

AndyAyersMS commented on src/coreclr/jit/lclmorph.cpp:
```diff
@@ -1353,6 +1353,77 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+                assert(TopValue(0).Node() == node->gtGetOp2());
+
+                // We only expect TopValue(1).IsAddress() to be true for stack allocated arrays
+                if (node->gtGetOp2()->IsCnsIntOrI() && TopValue(1).IsAddress())
```
I think this is too conservative. `GT_INDEX_ADDR` should never expose its op1 as it implies a bounds check.

Take a look at `IDEAEncryption:de_key_idea`: the array ends up exposed which blocks propagation of the length, which leads to regressions.

hez2010 replied:
I'm planning to move this to the global morph.

AndyAyersMS replied:
How is that going to help? If the array gets marked as exposed during local morph then the JIT will be forced to make very pessimistic assumptions about the contents.

We need to get the same sort of guarantees for a stack array that we have for a heap array. The codegen touching the array should be similar or better (we may see some extra prolog code to zero the array, so some size regressions are expected). So we need to dig into cases where it's not that way yet.

If all array indices are known we might expect to see promotion of array elements. So it would be good to verify that this works in simple cases.

If all array indices are known good things happen, but if any index is not known then things do not go as well. For example:

```C#
    public static int f()
    {
        int[] a = [1, 2];
        int i = 0;
        int j = 1;

        return a[i] + a[j];
    }
```
Currently this gives:
```asm
; Method f():int (FullOpts)
G_M32463_IG01:  ;; offset=0x0000
       sub      rsp, 56
       vxorps   xmm4, xmm4, xmm4
       vmovdqa  xmmword ptr [rsp+0x20], xmm4
       xor      eax, eax
       mov      qword ptr [rsp+0x30], rax
						;; size=21 bbWeight=1 PerfScore 3.83

G_M32463_IG02:  ;; offset=0x0015
       mov      rax, 0x7FF7E0BEECC8      ; int[]
       mov      qword ptr [rsp+0x20], rax
       mov      dword ptr [rsp+0x28], 2
       mov      dword ptr [rsp+0x30], 1
       mov      dword ptr [rsp+0x34], 2
       lea      rax, [rsp+0x20]
       mov      eax, dword ptr [rax+0x08]
       test     eax, eax
       je       SHORT G_M32463_IG04
       mov      ecx, dword ptr [rsp+0x30]
       cmp      eax, 1
       jbe      SHORT G_M32463_IG04
       mov      eax, ecx
       add      eax, dword ptr [rsp+0x34]
						;; size=66 bbWeight=1 PerfScore 12.50

G_M32463_IG03:  ;; offset=0x0057
       add      rsp, 56
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25

G_M32463_IG04:  ;; offset=0x005C
       call     CORINFO_HELP_RNGCHKFAIL
       int3     
						;; size=6 bbWeight=0 PerfScore 0.00
; Total bytes of code: 98
```

AndyAyersMS replied:
Seems like local morph doesn't really have a concept for an indexable local. I think we can get past this without adding a bunch of new concepts, but we'll see.

AndyAyersMS replied:
Not going to be quite as easy as I'd hoped. The code you have handles the "known index" by flattening it to a GT_LCL_FLD. But if the index is not known, it seems like we have two choices:
1. Keep the current INDEX_ADDR(GT_LCL_VAR_ADDR, index) and teach lcl morph, promotion, etc that there is now anther case besides ret buffs where GT_LCL_ADDR does not imply exposure; or
2. Introduce new concepts like GT_LCL_VAR_INDEX and STORE_GT_LCL_VAR_INDEX to encapsulate the entirety of the access and figure out how long those need to exist and how well they mesh with our other array access operations and such. Seems like not long after this we will want GT_LCL_VAR_INDEX_FLD to handle arrays of structs.

(1) is less disruptive but adds annoying context to GT_LCL_VAR_ADDR. We can annotate the nodes that happen under INDEX_ADDR and also mark the local in some way. That's the path I started down, so perhaps it's worth hacking it through to the "end" just to see how much else is impacted.

AndyAyersMS replied:
One snag with (1) is not only that the LCL_VAR_ADDR is contextual, but also that the necessary context can be two hops away, eg we need to differentiate IND(INDEX_ADDR(LCL_VAR_ADDR, ...)) from STOREIND(INDEX_ADDR(LCL_VAR_ADDR, ...)).

jakobbotsch replied:
It seem unlikely to me we'll be able to get very far trying to handle unknown indices -- no passes will be able to reason about these loads or stores, so it seems we will effectively have to be as conservative as if we had exposed it.

Is the goal to allow at least reasoning about the length field even when the elements accesses are unknown? I wonder if we could somehow split the array into an exposed and an unexposed part during stack allocation.

AndyAyersMS replied:
The goal is to not end up worse off if we stack allocate an array... if we don't stack allocate the above code is:
```asm
; Method A:f():int (FullOpts)
G_M32463_IG01:  ;; offset=0x0000
       sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25

G_M32463_IG02:  ;; offset=0x0004
       mov      rcx, 0x7FF7E6E6ECC8      ; int[]
       mov      edx, 2
       call     CORINFO_HELP_NEWARR_1_VC
       mov      rcx, 0x200000001
       mov      qword ptr [rax+0x10], rcx
       mov      ecx, dword ptr [rax+0x10]
       add      ecx, dword ptr [rax+0x14]
       mov      eax, ecx
						;; size=42 bbWeight=1 PerfScore 8.00

G_M32463_IG03:  ;; offset=0x002E
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 51
```
So a bit more efficient initialization and no bounds checks, and if we knew the array was thread private, perhaps the whole thing would optimize away.

Splitting might be an option, or (perhaps similar) we could model the on-stack array as span.

hez2010 replied:
With
```cs
var arr = new int[5];
arr[0] = 1; arr[1] = 2; arr[2] = 3; arr[3] = 4; arr[4] = 5;
Console.WriteLine(arr.Length);
Console.WriteLine(arr[0] + arr[1]);
```

We get 
```asm
G_M27646_IG01:  ;; offset=0x0000
       sub      rsp, 40
                                                ;; size=4 bbWeight=1 PerfScore 0.25
G_M27646_IG02:  ;; offset=0x0004
       mov      ecx, 5
       call     [System.Console:WriteLine(int)]
       mov      ecx, 3
       call     [System.Console:WriteLine(int)]
       nop
                                                ;; size=23 bbWeight=1 PerfScore 6.75
G_M27646_IG03:  ;; offset=0x001B
       add      rsp, 40
       ret
                                                ;; size=5 bbWeight=1 PerfScore 1.25
```

but with
```cs
var arr = new [] { 1, 2, 3, 4, 5 };
Console.WriteLine(arr.Length);
Console.WriteLine(arr[0] + arr[1]);
```

we get
```asm
G_M27646_IG01:  ;; offset=0x0000
       push     rsi
       push     rbx
       sub      rsp, 40
                                                ;; size=6 bbWeight=1 PerfScore 2.25
G_M27646_IG02:  ;; offset=0x0006
       mov      rcx, 0x1D38EBA2988      ; const ptr
       mov      ebx, dword ptr [rcx]
       mov      rcx, 0x1D38EBA298C
       mov      esi, dword ptr [rcx]
       mov      ecx, 5
       call     [System.Console:WriteLine(int)]
       lea      ecx, [rbx+rsi]
       call     [System.Console:WriteLine(int)]
       nop
                                                ;; size=45 bbWeight=1 PerfScore 11.50
G_M27646_IG03:  ;; offset=0x0033
       add      rsp, 40
       pop      rbx
       pop      rsi
       ret
                                                ;; size=7 bbWeight=1 PerfScore 2.25
```

AndyAyersMS replied:
I think a good goal here is to minimize the possible downside and not yet worry about maximizing the possible upside. So the comparison should be against the heap allocated code.

> this trivial case can be resolved by handling `INDEX_ADDR` and `ARR_LENGTH` in global morph so that the assertion can be propagated.

It does not seem trivial to me. If the array is address exposed how can morph make any assertions about its contents? Maybe VN can reason about some of this, since it tries to model memory updates, but then it is easily foiled by writes or calls. I can see (perhaps) special casing the length or method table fields, since we know they can't be safely modified.

hez2010 replied:
With my latest commit now the codegen of your snippet becomes:
```asm
; Program:f():int (FullOpts)

G_M52058_IG01:  ;; offset=0x0000
       sub      rsp, 24
       vxorps   xmm4, xmm4, xmm4
       vmovdqa  xmmword ptr [rsp], xmm4
       xor      eax, eax
       mov      qword ptr [rsp+0x10], rax
                                                ;; size=20 bbWeight=1 PerfScore 3.83
G_M52058_IG02:  ;; offset=0x0014
       mov      rax, 0x7FFD0E66ECC8      ; int[]
       mov      qword ptr [rsp], rax
       mov      dword ptr [rsp+0x08], 2
       mov      dword ptr [rsp+0x10], 1
       mov      dword ptr [rsp+0x14], 2
       mov      eax, dword ptr [rsp+0x10]
       add      eax, dword ptr [rsp+0x14]
                                                ;; size=46 bbWeight=1 PerfScore 7.25
G_M52058_IG03:  ;; offset=0x0042
       add      rsp, 24
       ret
                                                ;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code 71
```

which is better than before, and it's also better than heap allocated array.

hez2010 committed:
Introduce a flag for stack allocated arrays

hez2010 committed:
Handle in global morph

hez2010 committed:
Some special cases

hez2010 committed:
Avoid extensive remorph

hez2010 committed:
Oops

hez2010 commented:
@AndyAyersMS BTW we can mark `Array.Copy` and `SpanHelper.Memmove` as non-escaping to see if it can give us more opportunities.

AndyAyersMS commented:
> @AndyAyersMS BTW we can mark `Array.Copy` and `SpanHelper.Memmove` as non-escaping to see if it can give us more opportunities.

If we start passing stack allocated ref classes to callees we also have to fix the GC reporting for those callee arguments to be managed (not object) pointers (and transitively, fix reporting for any place those arguments can propagate, including possibly in the native parts of the runtime). So there is (perhaps considerable) extra work.

hez2010 commented:
@EgorBo and me discussed on discord that we can probe the size argument using value probing, so that for unknown sized arrays we can do "guarded stack allocation" in the future.

It would effectively replace
```cs
Span<int> arr = new int[size];
``` 
with
```cs
Span<int> arr;
if (size < 16)
{
    // dummy code below
    arr = stackalloc int[16];
    arr.Length = size;
}
else
{
    arr = new int[42];
}
```

AndyAyersMS commented:
@hez2010 I think we are getting close.

I think we still need to be thoughtful about the total size we want to allow here, especially so if the allocation site is conditional within the method (we likely should reduce the limit for objects too, `0x2000U` seems awfully large).

@jakobbotsch what are your thoughts about special-casing the bounds checks for stack allocated arrays to use the total allocation extent rather than looking at the length field?

I am tempted to try and unify stack allocated array checks with our existing bounds check logic by having the stack case invoke a "helper method" at the allocation site (which will lower to nothing or perhaps stores to the vtable and length field) so that the heap and stack cases are handled similarly.
Or maybe just a special flag on the existing new array helpers that says they are no-ops?

That way most of our existing optimizations should fire without any extra effort. Later if we want to take advantage of thread-privateness (say in VN) we should have the breadcrumbs we need to do this.

AndyAyersMS commented:
> I am tempted to try and unify stack allocated array checks with our existing bounds check logic by having the stack case invoke a "helper method" at the allocation site (which will lower to nothing or perhaps stores to the vtable and length field) so that the heap and stack cases are handled similarly.

I'm going to prototype this, since I think it's better than relying on the overall size or on being able to propagate the element type/length through memory. Aside from sharing the inference logic with heap arrays, it will also generalize better to variable-length or MD arrays...

AndyAyersMS committed:
Merge branch 'main' into value-array-stack-alloc

AndyAyersMS commented:
> > I am tempted to try and unify stack allocated array checks with our existing bounds check logic by having the stack case invoke a "helper method" at the allocation site (which will lower to nothing or perhaps stores to the vtable and length field) so that the heap and stack cases are handled similarly.
> 
> I'm going to prototype this, since I think it's better than relying on the overall size or on being able to propagate the element type/length through memory. Aside from sharing the inference logic with heap arrays, it will also generalize better to variable-length or MD arrays...

This seems simple enough, the allocation is already a helper call so we can mostly just leave it as is; however we likely also don't want to propagate the local address, so that dataflow on the array leads back to the helper, which means we lose the connection of the newarr temp and the local. We can add that as some kind of extra payload on the helper perhaps.

I also suspect what is in this PR doesn't always work right for R2R/NAOT, it looks like we embed the compile time handle into the array's method table slot, but then it's unlikely we ever look at this slot in cases where the array doesn't escape, so not sure if there's a test case that can be made to fail here.

AndyAyersMS commented:
If we defer "lowering" of these helpers until lower, we may miss out on forming address modes from the local directly... we'll have something like this during optimization
```
  t = newarr(len, type) [stack:V]
  ...
    = t[i]
```
and only later expose the assignment as to a local address:
```
    V[0] = method table
    V[8] = len 
    t = &V;
    ...
       = t[i]
```
so may end up needing an extra register perhaps.

AndyAyersMS committed:
Alternative take on array stack allocation

Leave the newarr helper call in place, and don't rewrite the uses to be
uses of the local. Remove special handling in local morph and morph.

Lower the helper to the proper stores later on.

Update a few utilities to understand array base addresses may not be TYP_REF.

AndyAyersMS committed:
basic VN support

AndyAyersMS commented:
I think this approach is looking pretty good. No changes now in morph or local morph or assertion prop. May not have the R2R embedding right yet either, but it may not actually matter until we try and support arrays with GC types.

We aren't as capable as optimizing the results as I'd like, but improving on that raises deeper questions of our aliasing model, and so likely should be deferred.

For example, liveness will mark all STOREINDs as modifying the GC heap, but now we have some that won't. Detecting that requires annotating the store with something like GTF_IND_TGT_NOT_HEAP, but I don't want to rely on that in liveness since it also can refer to pinned heap or unmanaged non-stack memory, so we probably want a different attribute like GTF_IND_TGT_STACK, and a separation of memory SSA tracking into heap, stack (byref), and unmanaged?

For now I have plumbed through some logic in VN to recognize these new local arrays can be treated via liberal VNs, this enables constant prop through known index elements as long as there is no ambiguous heap store in between. But without VN we don't have enough breadcrumbs to model these more aggressively.

I still need to look more closely at diffs between this and the above. SPMI doesn't show many, but has a few hundred missed contexts per collection, I suspect behavior may diverge there.

AndyAyersMS committed:
Merge branch 'main' into value-array-stack-alloc

AndyAyersMS committed:
restore complexity change

AndyAyersMS committed:
pass address of stack local to new helper

AndyAyersMS committed:
Merge branch 'main' into value-array-stack-alloc-alt

AndyAyersMS committed:
temp hack to boost SPMI coverage

AndyAyersMS committed:
avoid pessimizing tail calls. implement configurable size limit

AndyAyersMS committed:
add missing well known arg string

AndyAyersMS committed:
fix array length check

AndyAyersMS committed:
use ClrSafeInt directly

AndyAyersMS committed:
bypass for R2R for now, since it may inhibit prejitting

AndyAyersMS committed:
Merge branch 'main' into value-array-stack-alloc

AndyAyersMS committed:
Merge remote-tracking branch 'hez/value-array-stack-alloc' into value-array-stack-alloc

AndyAyersMS commented:
Merged in the changes from #111284.

AndyAyersMS commented:
@dotnet/jit-contrib PTAL
cc @jakobbotsch @hez2010 

SPMI will be fairly accurate here... kicked off another collection which should have even fewer misses. Code size increases but mostly from clr test where small fixed-sized arrays are unusually frequent.

* We currently limit array size to 528 bytes (so enough for a 512 byte array or a 128 int array). This can be altered by config.
* I have disabled this opt for R2R because it inhibits some cross-module inlines. If/when we can prove that the vtable is never looked at (or perhaps rely on a runtime helper call to initialize it), we can revisit.
* There is no "hotness" check so we will stack allocate arrays that are created in cold blocks. We should reconsider this since these arrays will frequently require prolog zeroing.
* There is some special handling in VN to always rely on liberal VNs, but it is still simplistic, eg for a local array `a` the array load below will not be recognized as a constant.
```C#
 a[0] = 3;
 f();
     = a[0];
```
* Arrays that are only stored to are not optimized away
* Other opts mostly work as they do now, as these arrays don't really look any different than heap arrays to the jit.

AndyAyersMS commented:
Seeing AVs in osx crossgen2, oddly in both base and diff jits:
```
[17:42:02] Invoking: C:\h\w\AAC9097D\p\superpmi.exe -a -v ewi -f C:\h\w\AAC9097D\t\tmpwd8c5ge0\libraries.crossgen2.osx.arm64.checked.mch_fail.mcl -details C:\h\w\AAC9097D\t\tmpwd8c5ge0\libraries.crossgen2.osx.arm64.checked.mch_details.csv -target arm64 -jitoption force JitEnableNoWayAssert=1 -jitoption force JitNoForceFallback=1 -jitoption force JitAlignLoops=0 -jit2option force JitEnableNoWayAssert=1 -jit2option force JitNoForceFallback=1 -jit2option force JitAlignLoops=0 -p -failureLimit 100 C:\h\w\AAC9097D\p\base\checked\clrjit_universal_arm64_x64.dll C:\h\w\AAC9097D\p\diff\checked\clrjit_universal_arm64_x64.dll C:\h\w\AAC9097D\w\9D38092F\e\artifacts\spmi\mch\cc0e7adf-e397-40b6-9d14-a7149815c991.osx.arm64\libraries.crossgen2.osx.arm64.checked.mch

[17:45:49] ERROR: Unexpected exception c0000005 was thrown.

[17:45:49] ERROR: Method 86217 of size 23 failed to load and compile correctly by JIT1 (C:\h\w\AAC9097D\p\base\checked\clrjit_universal_arm64_x64.dll).

[17:45:49] ERROR: Unexpected exception c0000005 was thrown.

[17:45:49] ERROR: Method 86217 of size 23 failed to load and compile correctly by JIT2 (C:\h\w\AAC9097D\p\diff\checked\clrjit_universal_arm64_x64.dll).

[17:45:49] ERROR: Unexpected exception c0000005 was thrown.
```
Will try and look at this locally

AndyAyersMS commented:
Failing here: `InitClass` map is null
```
CorInfoInitClassResult MethodContext::repInitClass(CORINFO_FIELD_HANDLE   field,
                                                   CORINFO_METHOD_HANDLE  method,
                                                   CORINFO_CONTEXT_HANDLE context)
{
    Agnostic_InitClass key;
    ZeroMemory(&key, sizeof(key)); // Zero key including any struct padding
    key.field   = CastHandle(field);
    key.method  = CastHandle(method);
    key.context = CastHandle(context);

    DWORD value = InitClass->Get(key);
```

hez2010 committed:
Merge branch 'main' into value-array-stack-alloc

hez2010 commented:
Merged the main branch to include #111555.

hez2010 commented:
Try a late dead store removal
@MihuBot

hez2010 commented:
@AndyAyersMS Just experimented a bit with late dead stores removal by unexposing locals in the liveness and repeating to convergence. 

Before adding late dead stores removal: [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=921915&view=ms.vss-build-web.run-extensions-tab)
After adding late dead stores removal: [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=922007&view=ms.vss-build-web.run-extensions-tab)

Relative diffs:
```
linux arm64: -26.588 bytes
linux x64: -27,113 bytes
osx arm64: -14,360 bytes
windows arm64: -25,528 bytes
windows x64: -26,628 bytes
```

Diffs look interesting but TP impacts are too high. Now I'm reverting the experiment commit.

AndyAyersMS committed:
Merge branch 'main' into value-array-stack-alloc

AndyAyersMS commented:
@jakobbotsch can you look this one over again?

jakobbotsch commented on src/coreclr/jit/gentree.h:
```diff
@@ -3586,7 +3586,7 @@ class SsaNumInfo final
 };
 
-// Common supertype of [STORE_]LCL_VAR, [STORE_]LCL_FLD, PHI_ARG, LCL_VAR_ADDR, LCL_FLD_ADDR.
+// Common supertype of [STORE_]LCL_VAR, [STORE_]LCL_FLD, PHI_ARG, LCL_ADDR, FIELD_ADDR.
```
```suggestion
// Common supertype of [STORE_]LCL_VAR, [STORE_]LCL_FLD, PHI_ARG, LCL_ADDR.
```

`FIELD_ADDR` does not derive from `GenTreeLclVarCommon`

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -2712,3 +2712,166 @@ bool Compiler::fgLateCastExpansionForCall(BasicBlock** pBlock, Statement* stmt,
+            return false;
+    }
+
+    if ((call->gtCallMoreFlags & GTF_CALL_M_STACK_ARRAY) == 0)
+    {
+        return false;
+    }
+
+    CallArg* const             stackLocalAddressArg  = call->gtArgs.FindWellKnownArg(WellKnownArg::StackArrayLocal);
```
Can we avoid introducing the flag by doing `stackLocalAddressArg != nullptr` instead?

jakobbotsch commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -13192,6 +13192,8 @@ const char* Compiler::gtGetWellKnownArgNameForArgMsg(WellKnownArg arg)
         case WellKnownArg::X86TailCallSpecialArg:
             return "tail call";
+        case WellKnownArg::StackArrayLocal:
+            return "stack array local";
```
Does this fit in the JITDUMP?

AndyAyersMS replied:
It did, but shortened it some.

jakobbotsch commented on src/coreclr/jit/objectalloc.cpp:
```diff
@@ -388,59 +396,72 @@ bool ObjectAllocator::MorphAllocObjNodes()
+                        case CORINFO_HELP_NEWARR_1_DIRECT:
+                        case CORINFO_HELP_NEWARR_1_ALIGN8:
+                        {
+                            if ((data->AsCall()->gtArgs.CountArgs() == 2) &&
+                                data->AsCall()->gtArgs.GetArgByIndex(1)->GetNode()->IsCnsIntOrI())
+                            {
+                                allocType = OAT_NEWARR;
+                            }
+                            break;
+                        }
+#ifdef FEATURE_READYTORUN
+                        case CORINFO_HELP_READYTORUN_NEWARR_1:
+                        {
+                            // Disable for R2R for now, since embedding the array type can block prejitting
+                            //
+                            if (!comp->opts.IsReadyToRun() && (data->AsCall()->gtArgs.CountArgs() == 1) &&
+                                data->AsCall()->gtArgs.GetArgByIndex(0)->GetNode()->IsCnsIntOrI())
```
I would suggest to change this to `CountUserArgs()` and `GetUserArgByIndex` to avoid potential issues with R2R indirection cells (probably no actual issue here since this runs before morph and helpers may not get the indirection cell arguments at all)

jakobbotsch commented on src/coreclr/jit/objectalloc.cpp:
```diff
@@ -454,51 +475,163 @@ bool ObjectAllocator::MorphAllocObjNodes()
+#ifdef FEATURE_READYTORUN
+                        if (comp->opts.IsReadyToRun() && data->IsHelperCall(comp, CORINFO_HELP_READYTORUN_NEWARR_1))
+                        {
+                            len = data->AsCall()->gtArgs.GetArgByIndex(0)->GetNode();
```
Dead code?

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -2712,3 +2712,166 @@ bool Compiler::fgLateCastExpansionForCall(BasicBlock** pBlock, Statement* stmt,
+    //
+    CORINFO_CLASS_HANDLE arrayHnd = (CORINFO_CLASS_HANDLE)call->compileTimeHelperArgumentHandle;
+
+    // Hack to reduce SPMI failures
```
Remove?

AndyAyersMS replied:
yep

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -2712,3 +2712,166 @@ bool Compiler::fgLateCastExpansionForCall(BasicBlock** pBlock, Statement* stmt,
+    }
+
+    CallArg* const             stackLocalAddressArg  = call->gtArgs.FindWellKnownArg(WellKnownArg::StackArrayLocal);
+    GenTreeLclVarCommon* const stackLocalAddressNode = stackLocalAddressArg->GetNode()->AsLclVarCommon();
+    const unsigned             lclNum                = stackLocalAddressNode->GetLclNum();
```
What guarantees that this `LCL_ADDR` makes it through the JIT? This looks pretty fragile, I do not think we should be making assumptions on the shape of this argument node.

jakobbotsch replied:
Seems like the code below can work just fine without making assumptions about the shape (potentially just optimizing for the `LCL_ADDR` shape)

AndyAyersMS replied:
So you are suggesting just using indirs to store the method table/length rather than fields? That is, transform
```
temp = newarray (mt; len, localarg)
```
into
```
temp = localarg
*temp = mt
*(temp + 8) = len
```
instead of what we're doing now?
```
*(&L) = mt
*(&L + 8) = len
temp = &L
```

jakobbotsch replied:
Yeah, that's what I was thinking.

(As a side note you are guaranteed after `gtSplitTree` that `localarg` is an invariant node, so you can clone it directly via `gtCloneExpr` without introducing any temp yourself)

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -2712,3 +2712,166 @@ bool Compiler::fgLateCastExpansionForCall(BasicBlock** pBlock, Statement* stmt,
+    gtSetStmtInfo(mtStmt);
+    fgSetStmtSeq(mtStmt);
+
+    // Initialize the array length.
+    //
+    GenTree* const   lengthArg    = call->gtArgs.GetArgByIndex(lengthArgIndex)->GetNode();
+    GenTree* const   lengthArgInt = fgOptimizeCast(gtNewCastNode(TYP_INT, lengthArg, false, TYP_INT));
+    GenTree* const   lengthStore = gtNewStoreLclFldNode(lclNum, TYP_INT, OFFSETOF__CORINFO_Array__length, lengthArgInt);
+    Statement* const lenStmt     = gtNewStmt(lengthStore);
+
+    fgInsertStmtBefore(block, newStmt, lenStmt);
+    gtSetStmtInfo(lenStmt);
+    fgSetStmtSeq(lenStmt);
+
+    // Replace call with &lclNum.
+    //
+    *callUse = stackLocalAddressNode;
+    DEBUG_DESTROY_NODE(call);
+
+    gtUpdateStmtSideEffects(stmt);
+    gtSetStmtInfo(stmt);
+    fgSetStmtSeq(stmt);
```
The insertions before `newStmt` is not what you want here. `newStmt` is the first new statement introduced by `gtSplitTree`, so this does not expand the allocation in its correct location. This looks like it should be inserting statements before `stmt`.

`stackLocalAddressNode` here refers to a node that `gtSplitTree` may have changed. That node should be obtained from the `CallArg` after the split has happened.

We also need to morph the block statements introduced by `gtSplitTree` since this is after morph. See e.g. the loop in `SplitTreesRandomly`. (I should clean up `gtSplitTree` to do this itself based on `fgGlobalMorphDone`.)

I would suggest the following sequence:

1. Split the tree
2. Morph block ops in the statements introduced by the splitting
3. Introduce the new statements before `stmt`
4. Replace the use
5. `fgMorphStmtBlockOps(stmt)` + `gtUpdateStmtSideEffects(stmt)`

jakobbotsch commented on src/coreclr/jit/objectalloc.h:
```diff
@@ -181,6 +254,11 @@ inline bool ObjectAllocator::CanAllocateLclVarOnStack(unsigned int         lclNu
         return false;
     }
```
It seems like we should have a limit on the aggregate stack allocated size. That's somewhat preexisting, but probably more important now.

jakobbotsch commented on src/coreclr/jit/valuenum.cpp:
```diff
@@ -6245,7 +6245,19 @@ void Compiler::fgValueNumberArrayElemLoad(GenTree* loadTree, VNFuncApp* addrFunc
     loadTree->gtVNPair.SetLiberal(loadValueVN);
-    loadTree->gtVNPair.SetConservative(vnStore->VNForExpr(compCurBB, loadType));
+
+    // If this is a local array, there are no asyncronous modifications, so we can set the
+    // conservative VN to the liberal VN.
+    //
+    VNFuncApp arrFn;
+    if (vnStore->IsVNNewLocalArr(arrVN, &arrFn))
+    {
+        loadTree->gtVNPair.SetConservative(loadValueVN);
+    }
+    else
+    {
+        loadTree->gtVNPair.SetConservative(vnStore->VNForExpr(compCurBB, loadType));
+    }
```
Does this actually show up as benefits?

AndyAyersMS replied:
In some limited cases, yes... eg we can const propagate through `a[2] = 1; y = a[2];`

AndyAyersMS committed:
revise helper expansion

AndyAyersMS committed:
revise remainder

AndyAyersMS commented:
@jakobbotsch think I've addressed most of the key points. Overall size limit will come in a future PR.

jakobbotsch commented:
@AndyAyersMS Did you push those changes?

AndyAyersMS commented:
Ah, I pushed to my fork, but ... this PR is not from my fork.

AndyAyersMS commented:
@jakobbotsch changes are there now

jakobbotsch commented on src/coreclr/jit/gentree.h:
```diff
@@ -4227,6 +4227,7 @@ enum GenTreeCallFlags : unsigned int
     GTF_CALL_M_CAST_CAN_BE_EXPANDED    = 0x04000000, // this cast (helper call) can be expanded if it's profitable. To be removed.
     GTF_CALL_M_CAST_OBJ_NONNULL        = 0x08000000, // if we expand this specific cast we don't need to check the input object for null
                                                      // NOTE: if needed, this flag can be removed, and we can introduce new _NONNUL cast helpers
+    GTF_CALL_M_STACK_ARRAY             = 0x10000000, // this call is a new array helper for a stack allocated array.
```
I suppose we weren't able to get rid of this since it's also used in VN.

jakobbotsch commented on src/coreclr/jit/valuenum.cpp:
```diff
@@ -14338,7 +14367,22 @@ bool Compiler::fgValueNumberHelperCall(GenTreeCall* call)
+                if (vnf == VNF_JitNewArr)
+                {
+                    vnf = VNF_JitNewLclArr;
+                    // modHeap = false;
+                }
+                else if (vnf == VNF_JitReadyToRunNewArr)
+                {
+                    vnf = VNF_JitReadyToRunNewLclArr;
+                    // modHeap = false;
```
Nit: commented code (feel free to remove as part of a follow up)

jakobbotsch commented on src/coreclr/jit/valuenum.cpp:
```diff
@@ -14338,7 +14367,22 @@ bool Compiler::fgValueNumberHelperCall(GenTreeCall* call)
                 }
             }
 
+            if (isAlloc && ((call->gtCallMoreFlags & GTF_CALL_M_STACK_ARRAY) != 0))
```
I suppose this could be changed to `FindWellKnownArg(StackArrayLocal) != nullptr` to get rid of the flag

jakobbotsch reviewed (approved):
LGTM

AndyAyersMS commented:
Thanks. I have more changes in this area so I can handle the last few bits in a subsequent PR.

AndyAyersMS commented:
@hez2010 thanks for all the work you did here.
